### DISCUSSION TITLE
How to use rehype-highlight in ReactMarkdown?

### DISCUSSION THREAD

**Discussion by dominicarrojado** [rehypejs/rehype#69] [Q&A] [answered]
Hopefully it's okay to post it here. I was trying to use rehype-highlight in ReactMarkdown. Btw my ReactMarkdown works just fine without plugin. Based on the doc, I should be able to use it like this:

```tsx
import ReactMarkdown from 'react-markdown'
import rehypeHighlight from 'rehype-highlight';

ReactDom.render(
  <ReactMarkdown rehypePlugins={[rehypeHighlight]}>
    This ~is not~ strikethrough, but ~~this is~~!
  </ReactMarkdown>,
  document.body
)
```

but I'm getting an error in my Next.js app:

```bash
Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Package subpath './lib/languages/arduino' is not defined by "exports" in /xxx/node_modules/highlight.js/package.json imported from /xxx/node_modules/lowlight/lib/common.js
```

I tried reproducing in a new Create React App project but get a different error:
```bash
./node_modules/hast-util-to-text/index.js 363:65
Module parse failed: Identifier directly after number (363:65)
File was processed with these loaders:
 * ./node_modules/react-scripts/node_modules/babel-loader/lib/index.js
You may need an additional loader to handle the result of these loaders.
|     //     break is the zero-width space character (U+200B), then the break is
|     //     removed, leaving behind the zero-width space.
>     if (lines[index].charCodeAt(lines[index].length - 1) === 0x20_0b
|     /* ZWSP */
|     || index < lines.length - 1 && lines[index + 1].charCodeAt(0) === 0x20_0b
```

**Comment by ChristianMurphy** [comment, marked as answer]
@dominicarrojado what versions of each library are you using?
Trying out an example in codesandbox https://codesandbox.io/s/react-markdown-rehype-highlight-32obr?file=/src/App.js
With
* `react-markdown@7.0.1`
* `rehype-highlight@5.0.0`
* `highlight.js@11.2.0`
* `react-scripts@4.0.3`

appears to work as expected https://codesandbox.io/s/react-markdown-rehype-highlight-32obr?file=/src/App.js.

---

For Next.js have you enabled ESM support? https://nextjs.org/blog/next-11-1#es-modules-support
And what versions of libraries are you using?

---

For CRA what version of libraries are you using? Have you customized the browserslist by any chance?

**Reply by dominicarrojado** [reply]
Hey @ChristianMurphy, thanks for the quick reply. I have the same version as you but I created the project via `yarn create react-app my-app --template typescript`. I didn't modified browserslist or any config.

For Next.js, I have enabled ESM support, but still get the same error. Here's the versions:
```json
  "dependencies": {
    "axios": "^0.21.2",
    "classnames": "^2.3.1",
    "date-fns": "^2.22.1",
    "gray-matter": "^4.0.3",
    "highlight.js": "^11.2.0",
    "moveto": "^1.8.2",
    "next": "11.1.1",
    "react": "17.0.2",
    "react-dom": "17.0.2",
    "react-lazyload": "^3.2.0",
    "react-markdown": "^7.0.1",
    "react-transition-group": "^4.4.2"
  },
  "devDependencies": {
    "@tailwindcss/typography": "^0.4.1",
    "@testing-library/jest-dom": "^5.14.1",
    "@testing-library/react": "^12.0.0",
    "@testing-library/react-hooks": "^7.0.1",
    "@types/axios": "^0.14.0",
    "@types/faker": "^5.5.7",
    "@types/jest": "^26.0.24",
    "@types/moveto": "^1.8.1",
    "@types/node": "^16.3.3",
    "@types/react": "17.0.14",
    "@types/react-lazyload": "^3.1.1",
    "@types/react-transition-group": "^4.4.2",
    "autoprefixer": "^10.3.1",
    "eslint": "7.31.0",
    "eslint-config-next": "11.0.1",
    "eslint-config-prettier": "^8.3.0",
    "faker": "^5.5.3",
    "gh-pages": "^3.2.3",
    "jest": "^27.0.6",
    "postcss": "^8.3.6",
    "tailwindcss": "^2.2.6",
    "typescript": "4.3.5"
  }
```

**Reply by ChristianMurphy** [reply]
@dominicarrojado could you share a runnable example of the issue you are seeing in a CodeSandbox?
I tried with CRA TypeScript as well and it works https://codesandbox.io/s/react-markdown-typescript-highlight-js-u9u9d?file=/src/App.tsx :thinking: 
They support a number of frameworks in CodeSandbox, including CRA, CRA+TS, and Next.js.

**Reply by dominicarrojado** [reply]
@ChristianMurphy I get the CRA error I got from my local if I try CodeSandbox for Next.js: https://codesandbox.io/s/interesting-fire-thu88?file=/pages/index.js

**Reply by ChristianMurphy** [reply]
It appears Next doesn't support numeric separators (https://v8.dev/features/numeric-separators) which is a standard feature in JavaScript.

A few things which may help:
* make sure you are running node 12 or higher (CodeSandbox's next template defaults to a server with node 10 IIRC)
* try customizing next's babel config https://nextjs.org/docs/advanced-features/customizing-babel-config to include https://babeljs.io/docs/en/babel-plugin-syntax-numeric-separator
* if the issue persists it may be a deeper issue in next's webpack configuration, raise a discussion/issue with Next to support numeric separators (https://github.com/vercel/next.js/discussions or https://github.com/vercel/next.js/issues)

related: https://github.com/syntax-tree/hast-util-to-text/issues/4

**Reply by dominicarrojado** [reply]
@ChristianMurphy hmm ok, thanks for checking. My node is on `v12.19.0` even previously, I did try customizing the babel config but I still get this error:
```bash
Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Package subpath './lib/languages/arduino' is not defined by "exports" in /xxx/node_modules/highlight.js/package.json imported from /xxx/node_modules/lowlight/lib/common.js
```

I'm still puzzled why on my local CRA TypeScript, I get this error which is similar to the CodeSandbox Next.js, cause you mentioned it could be next's webpack config but this is CRA just that it was installed in my local.
```bash
./node_modules/hast-util-to-text/index.js 363:65
Module parse failed: Identifier directly after number (363:65)
File was processed with these loaders:
 * ./node_modules/react-scripts/node_modules/babel-loader/lib/index.js
You may need an additional loader to handle the result of these loaders.
|     //     break is the zero-width space character (U+200B), then the break is
|     //     removed, leaving behind the zero-width space.
>     if (lines[index].charCodeAt(lines[index].length - 1) === 0x20_0b
|     /* ZWSP */
|     || index < lines.length - 1 && lines[index + 1].charCodeAt(0) === 0x20_0b
```

**Reply by ChristianMurphy** [reply]
> ```
> Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Package subpath './lib/languages/arduino' is not defined by "exports" in /xxx/node_modules/highlight.js/package.json imported from /xxx/node_modules/lowlight/lib/common.js
> ```

Looks like a bug in webpack handling highlight https://github.com/webpack/webpack/issues/13865

discovered here: https://github.com/highlightjs/highlight.js/issues/3223#issuecomment-853564400 and https://github.com/highlightjs/highlight.js/issues/3223#issuecomment-886237555

**Reply by dominicarrojado** [reply]
@ChristianMurphy yeah, I installed an older version `rehype-highlight@4.1.0` and it is working in Next.js.
